### ISSUE TITLE
Treat Location header similar to a link

### ISSUE THREAD

**Issue by basti1302** [basti1302/traverson#45]
A request might return a response with a location header. It would probably be quite useful if Traverson would be able to handle the location handler as a special kind of link (conceptually it is very similar to a link in the body) and could follow it.

This was brought up as an aside by @xogeny in #44.

**Comment by xogeny** [comment]
I agree. :smile: :+1:

There is a wrinkle to this.  A response that returns a `Location` header may also return a body.  I've seen it said that it is good practice to "pre-fetch" the contents of the `Location` header and include them in the body (e.g., after a `POST`) to save the client the step of doing a `GET`.  As such, I think it would be good to give the client the option to still use the contents of the body.  I suspect in practice that would mean a special alternative to `Follow` like `FollowLocation`.  That way, the client has the ability to deal with either the body or the `Location` URL.

**Comment by basti1302** [comment]
I was more thinking about a resource in the middle of the traversal, not at the end. For these resources, it is opaque to the client (to the code using Traverson) what happens. The client never sees the resource, nor the headers. So if we allow something like 

```
traverson
.from('http://api.example.com')
.json()
.follow('link1', 'link2', '$$header.location$$', 'link3')
.getResource(...);
```

This traversal would follow two plain vanilla links, then go to the URL of the location header of the resource behind link2 and then follow another plain link from there. This would also work with '$$header.location$$' being the last link-rel (or pseudo link-rel) to follow. Of course I'm making up the "syntax" for this use case as I go...

I'm not sure I fully follow your argument of giving the client both the last body before following the location header or the resource at the URL of the location header or even both. I would have guessed the client wants either or?

**Comment by xogeny** [comment]
OK, I see what you mean.  Let me show you want I had in mind.  It doesn't involve introducing any special "identifiers" for the follow.  The case you wrote could be written as:

```
traverson
.from('http://api.example.com')
.json()
.follow('link1', 'link2')
.followLocation()
.follow('link3')
.getResource(...);
```

The idea is that when it processes the `followLocation`, it uses the location header.  I prefer this more explicit version because it is distinct from `follow` and the link relations.  I just imagine a user coming along and reading code like `$$header.location$$` and thinking to himself "There's no relation called $$header.location$$ in this response, what is this doing?".  I know you weren't proposing this specific syntax, but the point stands regardless of what syntax you use.

My point about the body is that **if** a client knows that the API will return a body that contains the same content that a `GET(location.header)` would return, then they can just write the same request as:

```
traverson
.from('http://api.example.com')
.json()
.follow('link1', 'link2', 'link3')
.getResource(...);
```

*i.e.,* they can ignore the `Location` header completely.  Again, I like the explicit indicators here so we know which is being used.  There is no magic.  Another place you see this in HAL is when you get a response like this:

```
{
  "_links": {
    ...
    "next": { "href": "/one/path/to/next", "title": "Next item" },
    ...
  },
  "_embedded": {
    "next": {
      "x": 5,
      "_links": {
        "self": { "href": "/another/path/to/next" },
      },
    },
  },
}

What happens when you do:

```
traverson
.from('http://api.example.com')
.json()
.follow('next')
.getResource(...);
```

Do we get what is at `/one/path/to/next` or do we get the contents of the embedded resource (*e.g.,* where `x` is `5`)?  Note that the resource we fetch from `/one/path/to/next` might not even include `x` (per the HAL specification!).

I don't like this ambiguity either.  IIRC, `traverson` prefers the embedded over the link (but I could be wrong).  For me, I would prefer an explicit approach, *i.e.,*

```
traverson
.from('http://api.example.com')
.json()
.followEmbedded('next')
.getResource(...);
```

Otherwise, I don't really know what `traverson` is doing here.

But these are just my preferences.

**Comment by basti1302** [comment]
Wow, wall of text :-)  Let's go through this one by one.

## Implicit vs. Explicit

In general: Making things more explicit instead in favor of Traverson doing some implicit, intransparent magic: +1

## API Discussion

Regarding your code
```
.follow('link1', 'link2')
.followLocation()
.follow('link3')
```
I really like this approach for the API of this! A downside is that it implies different behaviour regarding the method follow from current versions of Traverson. Given the example above, one would need have `.follow('link1).follow('link2')` to result in a traversal via `link1, link2`. Currently, a subsequent call to `follow` overwrites  earlier calls, that is, the code above result in a traversal via `link2` only. Since there was no reason to call follow multiple times until now, this is probably not a big issue. One question that might play into this: Can we imagine a lot more special cases similar to `followLocation` in the future? Those would all be different methods then. 

## HAL

Regarding the comment on HAL and embedded versus linked resources: I think I do not like the explicit `followEmbedded` here as much as in the location header case. From what I understand when reading the HAL spec's section about the [Hypertext Cache Pattern](https://tools.ietf.org/html/draft-kelly-json-hal-07#section-8.3) and also the section about [embedded](https://tools.ietf.org/html/draft-kelly-json-hal-07#section-4.1.2), both the resource at `/one/path/to/next` and the embedded `next` resource (and thus also the resource at `/another/path/to/next`) ought to be the **same** resource. It is If not, the server is misbehaving. It might be a partial or inconsistent view, but it is still the same resource.

Currently, Traverson would prefer the linked resource over the embedded resource in the API. This can be configured with the [preferEmbeddedResources](https://github.com/basti1302/traverson/blob/master/api.markdown#builder-preferEmbeddedResources). Maybe both (the default and the config method) could be documented in a more prominent place but it is specified and not subject to arbitrary behaviour.

**Comment by basti1302** [comment]
This has been implemented in branch [follow-location-header-45](https://github.com/basti1302/traverson/tree/follow-location-header-45), in particular commit https://github.com/basti1302/traverson/commit/0517c693ad22543f93b45e15b3e6c4cb1ba4139d

[test/localhost.js](https://github.com/basti1302/traverson/blob/follow-location-header-45/test/localhost.js#L184-L199) has a showcase for the new feature.

Would you like to have a look before it is merged into master?

**Comment by basti1302** [comment]
This has landed in 3.0.0 /cc @xogeny
